Low power timer needs to be reset when setting time

TacoGrandeTX opened the pull request:
### Description

This PR addresses https://github.com/ARMmbed/mbed-os/issues/7804.

When set_time() is called with DEVICE_LPTICKER the timer needs to be reset also, otherwise
the old _start value is incorrect and this taints the value returned by Timer::read().

Using the issue's test program, the minutes and seconds are now properly tracked after calling set_time():

```
resetting time using hour 23...
Thu Aug 16 23:01:01 2018
Thu Aug 16 23:01:02 2018
Thu Aug 16 23:01:03 2018
Thu Aug 16 23:01:04 2018
Thu Aug 16 23:01:05 2018
Thu Aug 16 23:01:06 2018
Thu Aug 16 23:01:07 2018
Thu Aug 16 23:01:08 2018
Thu Aug 16 23:01:09 2018
Thu Aug 16 23:01:10 2018
resetting time using hour 23...
Thu Aug 16 23:01:01 2018
Thu Aug 16 23:01:02 2018
Thu Aug 16 23:01:03 2018
Thu Aug 16 23:01:04 2018
Thu Aug 16 23:01:05 2018
Thu Aug 16 23:01:06 2018
Thu Aug 16 23:01:07 2018
Thu Aug 16 23:01:08 2018
Thu Aug 16 23:01:09 2018
Thu Aug 16 23:01:10 2018
resetting time using hour 23...
Thu Aug 16 23:01:01 2018
Thu Aug 16 23:01:02 2018
Thu Aug 16 23:01:03 2018
```


### Pull request type
    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Breaking change

TacoGrandeTX committed:
Low power timer needs to be reset when setting time.
Reset _rtc_lp_timer only if DEVICE_LPTICKER is defined.

TacoGrandeTX commented:
Adding @kegilbert for comment/review

kegilbert reviewed (commented):
LGTM, needs someone closer to the lpticker source to sign off though. Added @c1728p9 . Feel free to add anyone else.

cmonr commented:
@NeilMacMullen Would you mind confirming that this fixes the issue?

0xc0170 commented:
@TacoGrandeTX Review travis failure, seems like this change fails to build

NeilMacMullen commented:
@cmonr  Unfortunately I'm on vacation for the next couple of weeks so can't confirm  for a while but I think the original issue report should make a reasonable unit test.

bulislaw reviewed (dismissed):
We need tests to confirm the fix.

RobMeades commented:
Can this fix be progressed please?  It's not good to have something as fundamental as `set_time()` broken for so long.

0xc0170 commented:
@TacoGrandeTX Can you please add a test case ?

TacoGrandeTX commented:
Neil MacMullen's test case is a good one:
```
#include "mbed.h"

int main()
{
    for (int hour = 23; hour >= 0; hour--)
    {
        for (int z = 0; z < 12; z++)
        {
            printf("resetting time using hour %d...\n", hour);

            struct tm t
            {
            };
            t.tm_year = 2018 - 1900;
            t.tm_mon = 8 - 1;
            t.tm_mday = 16;
            t.tm_hour = hour;
            t.tm_min = 1;
            t.tm_sec = 1;
            set_time(mktime(&t));

            for (int c = 0; c < 10; c++)
            {
                time_t seconds = time(NULL);
                printf("%s", ctime(&seconds));
                wait_ms(1000);
            }
        }
    }
    return 0;
}
```
Before (note how seconds are not reset):
```
resetting time using hour 23...
Thu Aug 16 23:01:01 2018
Thu Aug 16 23:01:02 2018
Thu Aug 16 23:01:03 2018
Thu Aug 16 23:01:04 2018
Thu Aug 16 23:01:05 2018
Thu Aug 16 23:01:06 2018
Thu Aug 16 23:01:07 2018
Thu Aug 16 23:01:08 2018
Thu Aug 16 23:01:09 2018
Thu Aug 16 23:01:10 2018
resetting time using hour 23...
Thu Aug 16 23:01:11 2018
Thu Aug 16 23:01:12 2018
Thu Aug 16 23:01:13 2018
Thu Aug 16 23:01:14 2018
```
Now (seconds are reset after calling set_time()):
```
resetting time using hour 23...
Thu Aug 16 23:01:01 2018
Thu Aug 16 23:01:02 2018
Thu Aug 16 23:01:03 2018
Thu Aug 16 23:01:04 2018
Thu Aug 16 23:01:05 2018
Thu Aug 16 23:01:06 2018
Thu Aug 16 23:01:07 2018
Thu Aug 16 23:01:08 2018
Thu Aug 16 23:01:09 2018
Thu Aug 16 23:01:10 2018
resetting time using hour 23...
Thu Aug 16 23:01:01 2018
Thu Aug 16 23:01:02 2018
Thu Aug 16 23:01:03 2018
``` 

Tested this on nRF52_DK and also MAX32620FTHR as there was a community post like RobMeades' about set_time() not working more than once on that target.

cmonr commented:
@TacoGrandeTX Something looks odd with this PR's history. Could you take a look at it?

0xc0170 commented:
@TacoGrandeTX What @bulislaw requested, we have hal tests for RTC. If new test case could be added to check to confirm the fix

TacoGrandeTX committed:
Add a protected HAL set_time() test

When DEVICE_LPTICKER is defined set_time() only works correctly on
the first call. This test calls set_time() twice and ensures the
time set by both calls is correct.  This test only runs if
DEVICE_RTC or DEVICE_LPTICKER is defined.

TacoGrandeTX commented:
I've just created https://github.com/ARMmbed/mbed-os/pull/8094 as a HAL test case.

0xc0170 commented:
> I've just created #8094 as a HAL test case.

To my understanding the test case should be part of this PR ? Otherwise test case PR cant be tested (fails). Can you add a test to this PR?

TacoGrandeTX commented:
@0xc0170 Understood - I've brought the HAL test into this PR and have closed https://github.com/ARMmbed/mbed-os/pull/8094.

mprse reviewed (changes requested):
I don't think that `TESTS/mbed_hal/rtc_time/main.cpp` is correct place to add test for `time()`.

We already have tests for `time()/set_time()` here:
https://github.com/ARMmbed/mbed-os/blob/master/TESTS/mbed_drivers/rtc/main.cpp

But currently the test is only valid when `time()` is based on `RTC`. This test should be expanded to handle also case when `time()` is based on `LPTIMER` (or additional/similar test should be added). 

The following test case can be modified to set time twice:
https://github.com/ARMmbed/mbed-os/blob/c356f1f9d45a366fa1a89025e39b3aed54860db8/TESTS/mbed_drivers/rtc/main.cpp#L423-L434

TacoGrandeTX commented:
@mprse - I originally thought about that but changed course for the better.  The request was to add a HAL test, not a driver test.  Most of `mbed_drivers/rtc` is checking that the RTC is hooked up correctly.  As you noted, `DEVICE_RTC` is not defined for all targets so adding tests for non-RTC targets in `mbed_drivers/rtc` would be confusing.  
@bulislaw - could we have your opinion?

mprse commented:
Maybe we should consider to move the test from drivers directory to platform. Looks like this is more suitable place.
I think that the test was created when time() was implemented based on rtc only. Now when it can be implemented based on rtc or lp timer maybe we should put existing test in platform/time_rtc and add similar test in platform/time_lptimer?

0xc0170 commented:
@mprse As this is a fix and adds a test case, and current discussion is about the refactor (should be done). Can this be done in two steps? First,  this goes in as a fix with a new test case, second clean up tests for rtc ?

0xc0170 commented:
@mprse If agree, please rereview (still requested changes here)

0xc0170 commented:
/morph build

cmonr commented:
It's a bit noisy, but I think the new test needs to be checked out.

cmonr commented:
Please take a look at the ARCH_PRO target. It seems like it doesn't like the tests-mbed_hal-rtc_time.test set_time twice test.

c1728p9 commented:
Adding @jeromecoutant as this may have an impact on ST targets.

c1728p9 commented on platform/mbed_rtc_time.cpp:
```diff
@@ -55,6 +55,7 @@ static time_t _rtc_lpticker_read(void)
 
 static void _rtc_lpticker_write(time_t t)
 {
+    _rtc_lp_timer->reset();
```
If the RTC is also being used for other parts of the system, such as the low power timer, this may interfere with it. One possible less intrusive alternative is to only adjust `_rtc_lp_base` by taking into account the current time. Something like this:
```
_rtc_lp_base = t - _rtc_lp_timer->read();
```

kjbracey replied:
The current version is fine, I think. It's only resetting the internal LowPowerTimer there. 

The prefixing of everything in the file with `_rtc` is confusing.

I'd also specifically suggest not having `_rtc_lp_base` be a `uint64_t` though. It's a `time_t`. Probably wasting a word of RAM and similar amount of ROM there, at least while `time_t` is 32-bit.

Been looking into C++11 chrono recently, so getting a bit picky about correctly-typed time. Having it be a time_t makes clear that it is the actual time the clock was last adjusted, which is potentially useful information, and I could imagine an API call to read that. Russ's proposed adjustment would lose that information.

jeromecoutant commented:
> Adding @jeromecoutant as this may have an impact on ST targets.

In fact, there is no ST impact for this fix as our targets supports RTC

Thx

cmonr commented:
/morph test

cmonr commented:
Still needs work.

kjbracey commented:
Test is failing on targets that have neither RTC nor LPTICKER. Didn't think there were any of those left.

The rest of the testcases in "mbed_hal/rtc_time" don't actually depend on anything in the HAL, afaict. Maybe the new test should have gone into "mbed_drivers/rtc", where the other `time()` tests are.

That is are conditional on `DEVICE_RTC` though - would need a little tweak to work with the LPTICKER version. 

However, I would really like the `time()` call to work on those platforms. It would be good to have an alternative implementation for RTOS using `Kernel::get_ms_count()`, which means it would.

Like events now do, it's better to ask the RTOS for the time it already has than to have a separate timer, so I'd have it in precedence order "RTC->RTOS->LPTICKER->nothing".

TacoGrandeTX commented:
@kjbracey-arm  
> The rest of the testcases in "mbed_hal/rtc_time" don't actually depend on anything in the HAL, afaict. Maybe the new test should have gone into "mbed_drivers/rtc", where the other time() tests are.
> 
> That is are conditional on DEVICE_RTC though - would need a little tweak to work with the LPTICKER version.
> 
Yes - we agreed that we would refactor this and RTC tests after this was merged (suggested far above by @0xc0170 ).  There is a push for this change from @RobMeades.

TacoGrandeTX commented:
@cmonr  Should time-related functions such as `time(NULL)` and `set_time()` be working on all Mbed targets?  If so - does RTC need to be added back to `targets.json` for the ARCH_PRO and REALTEK_RTL8195AM targets?

I found https://github.com/ARMmbed/mbed-os/issues/7380 and see the underlying dilemma in https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_rtc_time.cpp

As I check `targets.json` there are plenty of targets that don't have either RTC or LPTICKER defined.  Both these devices have a rtc_api.c so not understanding why RTC was removed here https://github.com/ARMmbed/mbed-os/commit/6d90c2716d64343bb34376a4a9effacc7a7fca99.  Presumably some update is required to each target's rtc_api.c?

cmonr commented:
@TacoGrandeTX I would think so, unless there are specific reasons that that should be disabled for those targets.

@ARMmbed/mbed-os-core Thoughts on https://github.com/ARMmbed/mbed-os/pull/7849#issuecomment-431183512 ?

If RTC is off, then how are the current tests being run against any device?

kegilbert commented:
@kjbracey-arm Has a good point, RTC->RTOS->LPTICKER->None sounds pretty solid to me.

Should be fairly straight forward to add that ordering here: https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_rtc_time.cpp#L25-L74

TacoGrandeTX commented:
@cmonr:  
> If RTC is off, then how are the current tests being run against any device?

As @kjbracey-arm pointed out, the tests in mbed_hal/rtc_time are only testing _rtc_maketime() and _rtc_localtime() which don't rely on having a running clock!

@kjbracey-arm: 
> Maybe the new test should have gone into "mbed_drivers/rtc", where the other time() tests are.

I think I'm going to have to do that.  Thanks for pointing out the proper location.  I'll try this again.

> However, I would really like the time() call to work on those platforms. It would be good to have an alternative implementation for RTOS using Kernel::get_ms_count(), which means it would.
> 
> Like events now do, it's better to ask the RTOS for the time it already has than to have a separate timer, so I'd have it in precedence order "RTC->RTOS->LPTICKER->nothing".
> 
OK - I'll check with @kegilbert about raising a new issue for this.

kegilbert commented:
@TacoGrandeTX Yeah a new issue would be perfect, sorry should've clarified that as an aside and not blocking for this PR.

TacoGrandeTX commented:
Met with @studavekar yesterday and he advised protecting the test with #if, so that's been done.  Performed a merge to pick up the template test changes so this should now be good to go.

studavekar commented:
We need to revisit https://github.com/ARMmbed/mbed-os/commit/6d90c2716d64343bb34376a4a9effacc7a7fca99 and check if we can enable RTC on disabled devices.

0xc0170 commented:
> We need to revisit 6d90c27 and check if we can enable RTC on disabled devices.

Who's doind this action? Is this blocking this PR?

studavekar commented:
> Who's doind this action? Is this blocking this PR?

enablement  RTC on disabled devices is not blocking this PR.

cmonr commented:
@studavekar Mind creating an issue? The comment made it sound like this needed to be addressed in this PR.

0xc0170 commented:
This is now entering CI

/morph build

cmonr commented:
**Note**: This PR is now a part of a rollup PR (#8675).

Jenkins CI export nodes experienced many drops throughout the day causing false failures. In an attempt to get those PRs through CI, while keeping CI load low, several PRs have been bundled into a single rollup PR.

_If any more commits are made in this PR, this PR will remain open and have to go through CI on its own._

0xc0170 commented:
/morph test
/morph export-build
